src/cdk/scrolling/viewport-ruler.ts

```diff
@@ -30,25 +30,35 @@ export class ViewportRuler implements OnDestroy {
   private _viewportSize: {width: number; height: number};
 
   /** Stream of viewport change events. */
-  private _change: Observable<Event>;
+  private _change = new Subject<Event>();
 
-  /** Subscription to streams that invalidate the cached viewport dimensions. */
-  private _invalidateCache: Subscription;
+  /** Event listener that will be used to handle the viewport change events. */
+  private _changeListener = (event: Event) => {
+    this._change.next(event);
+  }
 
   constructor(private _platform: Platform, ngZone: NgZone) {
     ngZone.runOutsideAngular(() => {
-      this._change = _platform.isBrowser ?
-          merge(fromEvent(window, 'resize'), fromEvent(window, 'orientationchange')) :
-          observableOf();
-
-      // Note that we need to do the subscription inside `runOutsideAngular`
-      // since subscribing is what causes the event listener to be added.
-      this._invalidateCache = this.change().subscribe(() => this._updateViewportSize());
+      if (_platform.isBrowser) {
+        // Note that bind the events ourselves, rather than going through something like RxJS's
+        // `fromEvent` so that we can ensure that they're bound outside of the NgZone.
+        window.addEventListener('resize', this._changeListener);
+        window.addEventListener('orientationchange', this._changeListener);
+      }
+
+      // We don't need to keep track of the subscription,
+      // because we complete the `change` stream on destroy.
+      this.change().subscribe(() => this._updateViewportSize());
```

Comment:
Here I would change a little bit the piece code, I would do something like:

```ts
    this._change = _platform.isBrowser ?
      merge(fromEvent(window, 'resize'), fromEvent(window, 'orientationchange'))
        .pipe(runOutsideZone(ngZone)) :
      observableOf();

```
where:

```ts
/** Runs this observable outside the zone */
function runOutsideZone<T>(zone: NgZone): any {
  return (source: Observable<T>): Observable<T> => {
      return new Observable(observer => {
          return zone.runOutsideAngular<Subscription>(() => source.subscribe(observer));
      });
  };
}
```

Seems cleaner and more understandable IMO, also is less code.